Content.Shared/Xenoarchaeology/Artifact/XAT/BaseXATSystem.cs

```diff
@@ -0,0 +1,86 @@
+using Content.Shared.Xenoarchaeology.Artifact.Components;
+using Robust.Shared.Timing;
+
+namespace Content.Shared.Xenoarchaeology.Artifact.XAT;
+
+public abstract class BaseXATSystem<T> : EntitySystem where T : Component
+{
+    [Dependency] protected readonly IGameTiming Timing = default!;
+    [Dependency] protected readonly SharedXenoArtifactSystem XenoArtifact = default!;
+
+    protected EntityQuery<XenoArtifactComponent> _xenoArtifactQuery;
+    private EntityQuery<XenoArtifactUnlockingComponent> _unlockingQuery;
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        _xenoArtifactQuery = GetEntityQuery<XenoArtifactComponent>();
+        _unlockingQuery = GetEntityQuery<XenoArtifactUnlockingComponent>();
+    }
+
+    protected void XATSubscribeDirectEvent<TEvent>(XATEventHandler<TEvent> eventHandler) where TEvent : notnull
+    {
+        SubscribeLocalEvent<T, XenoArchNodeRelayedEvent<TEvent>>((uid, component, args) =>
+        {
+            var nodeComp = Comp<XenoArtifactNodeComponent>(uid);
+
+            if (!CanTrigger(args.Artifact, (uid, nodeComp)))
+                return;
+
+            var node = new Entity<T, XenoArtifactNodeComponent>(uid, component, nodeComp);
+            eventHandler.Invoke(args.Artifact, node, ref args.Args);
+        });
+    }
+
+    protected delegate void XATEventHandler<TEvent>(Entity<XenoArtifactComponent> artifact, Entity<T, XenoArtifactNodeComponent> node, ref TEvent args)
+        where TEvent : notnull;
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        // TODO: need a pre-function so we can initialize queries
+        // TODO: add a way to defer triggering artifacts to the end of the Update loop
+
+        var query = EntityQueryEnumerator<T, XenoArtifactNodeComponent>();
+        while (query.MoveNext(out var uid, out var comp, out var node))
```

Comment:
need to add flag for skipping all of this if system does not use UpdateXAT at all - maybe thats what Emo meant in todo above? then no need to fill query either...

Reply:
fixed by creating separate abstract class for update-related types